Preparation work for NX migration

atarix83 opened the pull request:
## References  
* Related to architectural cleanup and modularization  https://wiki.lyrasis.org/display/DSPACE/DSpace+Angular+%3A+library-based+architecture+proposal
* This will replaced the first migration PR https://github.com/DSpace/dspace-angular/pull/4019

## Description  
This PR refactors and reorganizes the codebase to support a future migration to NX. The migration itself is **not** included in this PR—this is a **preparatory step** to make the transition smoother and more maintainable.

## Instructions for Reviewers  

This PR focuses on architectural improvements and code cleanup. The goal is to isolate core logic and reduce inter-module dependencies, paving the way for a modular NX-based structure.
Although the PR may appear large, most of the changes are due to updated import paths for core, config, and utils, which have been simplified to reflect how these modules will be used as libraries after the NX migration.

List of changes in this PR:  
* Isolated `core`, `config`, and `utils` by removing dependencies between those "modules" and other parts of the application  
* Moved `notifications` and `correlationId` state into the `core` state  
* Reorganized classes and files to reflect a clearer modular structure  
* Simplified import paths for `core`, `config`, and `utils` to reflect how they will be used as libraries after the NX migration  
* General code cleanup and removal of circular dependencies  

**How to test this PR:**  
* Ensure the application builds and runs correctly  
* Verify that `notifications` and `correlationId` functionalities are still working as expected  
* Check that there are no unnecessary dependencies between modules  
* Evaluate how well the different parts of the application have been isolated
* Identify any remaining classes or logic that should be moved into core, config, or utils to complete the modularization

## Checklist
- [x] My PR is **created against the `main` branch** of code (unless it is a backport or is fixing an issue specific to an older branch).
- [x] My PR is **small in size** (e.g. less than 1,000 lines of code, not including comments & specs/tests), or I have provided reasons as to why that's not possible.
- [x] My PR **passes [ESLint](https://eslint.org/)** validation using `npm run lint`
- [x] My PR **doesn't introduce circular dependencies** (verified via `npm run check-circ-deps`)
- [x] My PR **includes [TypeDoc](https://typedoc.org/) comments** for _all new (or modified) public methods and classes_. It also includes TypeDoc for large or complex private methods.
- [x] My PR **passes all specs/tests and includes new/updated specs or tests** based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] My PR **aligns with [Accessibility guidelines](https://wiki.lyrasis.org/display/DSDOC8x/Accessibility)** if it makes changes to the user interface.
- [x] My PR **uses i18n (internationalization) keys** instead of hardcoded English text, to allow for translations.
- [x] My PR **includes details on how to test it**. I've provided clear instructions to reviewers on how to successfully test this fix or feature.
- [x] If my PR includes new libraries/dependencies (in `package.json`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR includes new features or configurations, I've provided basic technical documentation in the PR itself.
- [x] If my PR fixes an issue ticket, I've [linked them together](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

tdonohue commented on tsconfig.json:
```diff
@@ -40,7 +40,12 @@
+    "paths": {
+      "@dspace/core/*": ["src/app/core/*"],
+      "@dspace/config/*": ["src/config/*"],
+      "@dspace/shared/utils/*": ["src/app/utils/*"]
```
@atarix83 : I gave this PR a quick review today.  Overall, it looks good to me.  I like that these paths simplify our imports (removing a lot of the messy relative paths (i.e. paths starting with things like `../../../../../`).

I think this change us useful regardless of whether we move to `Nx`.  But, if we don't move to `Nx`, it might be better to rename some of these paths to match with our directory structure like:
```
"@dspace/app/core/*": ["src/app/core/*"],
"@dspace/config/*": ["src/config/*"],
"@dspace/app/utils/*": ["src/app/utils/*"]
```

I.e. make `@dspace` just a stand-in for the `src/` folder.   This would make it easier on developers to know where to find the impored files without having to remember that `@dspace/core` is under `src/app/core`, but `@dspace/config` is one level up at `src/config`.

In general though, I like the addition of these paths whether we move to `Nx` or not.

(NOTE: I haven't had a chance to test this PR or do a thorough review.  I just reviewed the concepts and skimmed the code.)

ybnd replied:
I strongly agree with @tdonohue about limiting the paths to `@dspace` -> `src/*`.

Right now you've got a mix of `@dspace/...`, `src/..` and `../...` style import paths
- Before this PR our paths were already inconsistent, now even more so
- Some of the remaining `src/...` and `../...` style import paths _can_ already be cleanly replaced with `@dspace/...`, but were presumably missed?

Most of the conflicts this PR will cause would probably be mismatching import paths. To make resolving these conflicts feasible for other PRs and forks down the line we need some migration script IMO.
- I'm interested in taking a stab at it if that would be helpful

The "simplified path rule" will give us the opportunity to replace all `src` import paths at once and make everything consistent.
- This should be much easier to do with a script than the current situation
- We could look for a lint rule to enforce the `@dspace` prefix for 
- Any further module path changes will then become simple moves and the paths will remain consistent.

tdonohue replied:
In today's Developers Meeting, @atarix83 clarified why he specified the paths as:
```
"@dspace/core/*": ["src/app/core/*"],
"@dspace/config/*": ["src/config/*"],
"@dspace/shared/utils/*": ["src/app/utils/*"]
```

The goal of these paths is to prepare for different "libraries" or "modules".  So, the left side of the path is the future module name (core, config, shared/utils) while the right side is where that code *currently exists* in our codebase.  So, while these paths look confusing now, the purpose is to avoid having to change all the import statements *twice*, as @atarix83 is anticipating that these module directories will *move* to a new location once we modularize (regardless of whether we use Nx or another tool).

I can see that point, and I'm OK with this approach assuming that we are all in agreement on defining these named libraries/modules.  

I do agree that it could be nice to have a migration script though, as @ybnd notes above.

artlowel reviewed (commented):
Thanks @atarix83!

I agree that this is a good first step to modularize our code, regardless of whether we switch to NX

I've mainly worked on the code review so far, and would like to do some more testing to ensure that nothing breaks, but if nothing does, you'll get a +1 from me

ybnd reviewed (changes requested):
Relating to my comment above, there are some examples of import paths that can be changed to the new `@dspace/...` style but weren't:
(linking here instead of commenting inline, because there's too many changes for my browser to handle...)
- [src/app/core/submission/submission-response-parsing.service.ts](https://github.com/4Science/DSpace-angular/blob/cf4942ac8f108f7ace42eed7cd6d2ca3a0b5fbee/src/app/core/submission/submission-response-parsing.service.ts#L16-L24) -> `../` can be changed to `@dspace/core/`
- [src/app/home-page/home-page.component.ts](https://github.com/4Science/DSpace-angular/blob/cf4942ac8f108f7ace42eed7cd6d2ca3a0b5fbee/src/app/home-page/home-page.component.ts#L12-L15) -> `src/config/` can be changed to `@dspace/config/`

These are just a few examples, there are more cases.
With `@dspace -> src/*` as the only path it would be easier to make everything consistent at once.


Other than that the reorganization is great:
- The overall number of dependencies almost halved (checked with `npx madge`), so that's a big improvement in terms of 
  > Check that there are no unnecessary dependencies between modules
  > Evaluate how well the different parts of the application have been isolated
- Overall module sizes remain the same, so no regressions there 👌 
- Didn't encounter any other drawbacks

atarix83 commented:
> Relating to my comment above, there are some examples of import paths that can be changed to the new `@dspace/...` style but weren't: (linking here instead of commenting inline, because there's too many changes for my browser to handle...)
> 
>     * [src/app/core/submission/submission-response-parsing.service.ts](https://github.com/4Science/DSpace-angular/blob/cf4942ac8f108f7ace42eed7cd6d2ca3a0b5fbee/src/app/core/submission/submission-response-parsing.service.ts#L16-L24) -> `../` can be changed to `@dspace/core/`
> 
>     * [src/app/home-page/home-page.component.ts](https://github.com/4Science/DSpace-angular/blob/cf4942ac8f108f7ace42eed7cd6d2ca3a0b5fbee/src/app/home-page/home-page.component.ts#L12-L15) -> `src/config/` can be changed to `@dspace/config/`
> 
> 
> These are just a few examples, there are more cases. With `@dspace -> src/*` as the only path it would be easier to make everything consistent at once.
> 
> Other than that the reorganization is great:
> 
>     * The overall number of dependencies almost halved (checked with `npx madge`), so that's a big improvement in terms of
>       > Check that there are no unnecessary dependencies between modules
>       > Evaluate how well the different parts of the application have been isolated
> 
>     * Overall module sizes remain the same, so no regressions there 👌
> 
>     * Didn't encounter any other drawbacks

@ybnd 
I've fixed the import starting with src/* which pointed to config, core or utils. Please mind that i've fixed only the ones outside the core, config or utils folder, because within the same folder the imports with the relative path are correct

artlowel commented:
@atarix83 My feedback, as discussed in today's meeting:

- It's currently a bit arbitrary where guards, resolvers, and route paths are located. A few are in the core library, but most are still in app. I ran into a problem with this for `endUserAgreementGuard`, which is in the core library, but needs paths from `info-routes.ts`, which is in app.
	- In my opinion it makes the most sense to keep all guards, resolvers and path definitions in app for the moment, until we can split it up further. That way core contains things like models and services, i.e. the very low level stuff, while things that are related to actually rendering the web app stay in app for now. Eventually they can ofcourse move to a separate library.

- `src/app/core/auth/models/normalized-auth-status.model.ts` is empty 
- CorrelationIdService [contains a console.log](https://github.com/DSpace/dspace-angular/blob/00929e08d573812e6378cd182ec086f8ddd1e59d/src/app/core/correlation-id/correlation-id.service.ts#L94)

tdonohue reviewed (approved):
👍 Thanks @atarix83 !  I gave this a re-review today, and tested as much of the application as I could.  The code looks good and I couldn't find any issues.
